src/state/routing/useRouterTrade.ts

```diff
@@ -0,0 +1,147 @@
+  const userSlippageTolerance = useUserSlippageTolerance()
+  const [deadline] = useUserTransactionTTL()
+
+  const { isUninitialized, isLoading, isFetching, isError, data } = useGetQuoteQuery(
+    amountIn && currencyOut
+      ? {
+          tokenInAddress: amountIn.currency.wrapped.address,
+          tokenInChainId: amountIn.currency.chainId,
+          tokenOutAddress: currencyOut.wrapped.address,
+          tokenOutChainId: currencyOut.chainId,
+          amount: amountIn.quotient.toString(),
+          type: 'exactIn',
+          recipient: account ?? undefined,
+          slippageTolerance:
+            typeof userSlippageTolerance === 'string' ? userSlippageTolerance : userSlippageTolerance.toSignificant(),
+          deadline: deadline.toString(),
+        }
+      : skipToken,
+    { pollingInterval: ms`5m` }
+  )
+
+  const routes = useRoutes(
+    amountIn?.currency,
+    currencyOut,
+    // important to check `isUninitialized` as `skipToken` still returns cached data
+    isUninitialized ? data : undefined
+  )
+
+  // todo(judo): validate block number for freshness
+
+  return useMemo(() => {
+    if (!amountIn || !currencyOut) {
+      return {
+        state: V3TradeState.INVALID,
+        trade: null,
+      }
+    }
+
+    if (isLoading) {
+      return {
+        state: V3TradeState.LOADING,
+        trade: null,
+      }
+    }
+
+    const amountOut = currencyOut && data ? CurrencyAmount.fromRawAmount(currencyOut, data.quote) : undefined
+
+    if (isError || !amountOut || routes.length === 0) {
+      return {
+        state: V3TradeState.NO_ROUTE_FOUND,
+        trade: null,
+      }
+    }
+
+    const trade = Trade.createUncheckedTradeWithMultipleRoutes<Currency, Currency, TradeType.EXACT_INPUT>({
+      //todo(judo): inputAmount should be from API
+      routes,
+      tradeType: TradeType.EXACT_INPUT,
+    })
+
+    return {
+      state: isFetching ? V3TradeState.SYNCING : V3TradeState.VALID,
+      trade: trade,
+    }
+  }, [amountIn, currencyOut, isError, isLoading, routes, data, isFetching])
+}
+
+export function useRouterTradeExactOut(currencyIn?: Currency, amountOut?: CurrencyAmount<Currency>) {
+  const { account } = useActiveWeb3React()
+  const userSlippageTolerance = useUserSlippageTolerance()
+  const [deadline] = useUserTransactionTTL()
+
+  const { isUninitialized, isLoading, isFetching, isError, data } = useGetQuoteQuery(
+    amountOut && currencyIn
+      ? {
+          tokenInAddress: currencyIn.wrapped.address,
+          tokenInChainId: currencyIn.chainId,
+          tokenOutAddress: amountOut.currency.wrapped.address,
+          tokenOutChainId: amountOut.currency.chainId,
+          amount: amountOut.quotient.toString(),
+          type: 'exactOut',
+          recipient: account ?? undefined,
+          slippageTolerance:
+            typeof userSlippageTolerance === 'string' ? userSlippageTolerance : userSlippageTolerance.toSignificant(),
+          deadline: deadline.toString(),
```

Comment:
Since we need all 3 or none at all, will `deadline.toString()` work? Should we check all 3 are defined first?

Reply:
I actually check in routing/slice.ts that all 3 are defined. Here I'm just passing as much as possible to abstract away the API constraints.

Reply:
https://github.com/Uniswap/uniswap-interface/pull/2116/files#diff-c9fd69a3658e7e36f7ab6b62fd18589fed617dad15c3184bc7335b9e6ad7ee02R57